crates/types/src/lib.rs

```diff
@@ -1194,6 +1200,19 @@ impl ConstExpr {
     pub fn ops(&self) -> &[ConstOp] {
         &self.ops
     }
+
+    /// Is this ConstExpr a possibly zero integer value?
```

Comment:
Is this allowed to return false even when the value is actually zero? Jamey is adding support for the extended const proposal and it extends const expressions with addition, subtraction, and multiplication, so there will soon be many ways to evaluate to zero. Doesn’t mean we can’t land this PR now but we should be clear about the semantics of this method

Reply:
Clarified a bit -- renamed to `provably_nonzero_i32` (inverted from is-possibly-nonzero -- clearer this way IMHO) and commented more. It's always allowed to return `false` (can't prove anything), but if `true`, must actually be a case we know is always nonzero (currently only an i32.const with nonzero constant).

Reply:
I would suggest naming this with the opposite sense, like: `is_definitely_nonzero`, which returns true for `I32Const(x) if x != 0`, and may also return true if there are any other circumstances where it can prove this, but otherwise returns false. Then, extended constants or any other strange things will safely return false, meaning that you can't be sure your optimization is valid, and you'll only apply the optimization when it's definitely okay.

Further, when we follow Nick's suggestion for making the `ConstExpr` evaluator try to run even before an `Instance` is available, it will be clear how to invoke that to implement this. If it fails because it needs inputs which aren't known until instantiation, or it succeeds but returns 0, then we should return that the expression is not "definitely nonzero".

Reply:
Hah, raced; I'm glad we're on the same page.